modal/_utils/shell_utils.py

```diff
@@ -0,0 +1,113 @@
+# Copyright Modal Labs 2024
+import asyncio
+import contextlib
+import errno
+import os
+import select
+import sys
+from typing import Callable, Coroutine, Optional
+
+import rich.status
+
+from modal._pty import raw_terminal, set_nonblocking
+from modal.exception import ExecutionError, InteractiveTimeoutError
+
+from .async_utils import TaskContext, asyncify
+
+
+def write_to_fd(fd: int, data: bytes):
+    loop = asyncio.get_event_loop()
+    future = loop.create_future()
+
+    def try_write():
+        try:
+            nbytes = os.write(fd, data)
+            loop.remove_writer(fd)
+            future.set_result(nbytes)
+        except OSError as e:
+            if e.errno != errno.EAGAIN:
+                future.set_exception(e)
+                raise
+
+    loop.add_writer(fd, try_write)
+    return future
+
+
+@contextlib.asynccontextmanager
+async def stream_stdin(handle_input: Callable[[bytes, int], Coroutine], use_raw_terminal=False):
+    """Stream from terminal stdin to the handle_input provided by the method"""
+    quit_pipe_read, quit_pipe_write = os.pipe()
+
+    set_nonblocking(sys.stdin.fileno())
+
+    @asyncify
+    def _read_stdin() -> Optional[bytes]:
+        nonlocal quit_pipe_read
+        # TODO: Windows support.
+        (readable, _, _) = select.select([sys.stdin.buffer, quit_pipe_read], [], [], 5)
+        if quit_pipe_read in readable:
+            return None
+        if sys.stdin.buffer in readable:
+            return sys.stdin.buffer.read()
+        # we had 5 seconds of no input. send an empty string as a "heartbeat" to the server.
+        return b""
+
+    async def _write():
+        message_index = 1
+        while True:
+            data = await _read_stdin()
+            if data is None:
+                return
+
+            await handle_input(data, message_index)
+
+            message_index += 1
+
+    write_task = asyncio.create_task(_write())
+
+    if use_raw_terminal:
+        with raw_terminal():
+            yield
+    else:
+        yield
+    os.write(quit_pipe_write, b"\n")
+    write_task.cancel()
+
+
+async def connect_to_terminal(
+    # Handles data read from stdin.
+    handle_stdin: Callable[[bytes, int], Coroutine],
+    # Creates a coroutine that streams data to stdout/stderr.
+    stream_stdio: Callable[[asyncio.Event], Coroutine[None, None, int]],
+    pty: bool = False,
+    connecting_status: Optional[rich.status.Status] = None,
+):
+    """
+    Connect to the current terminal by streaming data from terminal's stdin to the running process
+    and streaming output from running process into terminal's stdout.
+
+    If connecting_status is given, this function will stop the status spinner upon connection or error.
+    """
+
```

Comment:
this function's code was originally inside connect_to_exec. Broke it into its own function that takes two `functions` to handle `stdin` and `stdout/stderr`